### PR TITLE
[Qwen3] StateDictAdapter support for MoE model

### PR DESCRIPTION
**Pull request by shuhuayu**
Reused StateDictAdapter support for DeepSeek V3 model to implement Qwen3 StateDictAdapter. Updated a checkpoint loading api to support distributed huggingface checkpoint loading when unpicklable objects exist.

### CODE REVIEW

File: torchtitan/experiments/qwen3/model/state_dict_adapter.py [hunk 1]
```diff
@@ -29,58 +31,180 @@ def __init__(self, model_args: Qwen3ModelArgs, hf_assets_path: str | None):
...
-        return hf_state_dict
+        # Prepare for dequantization
+        hf_state_dict_with_scale_inv = self._add_quantization_scale_inv_tensors(
+            hf_state_dict
+        )
+        return hf_state_dict_with_scale_inv
 
     def from_hf(self, hf_state_dict: dict[str, Any]) -> dict[str, Any]:
+        """
+        1. When loading from HF checkpoint, dequantize the weights from float8 to float32.
```
Comment (wwwjn) [inline torchtitan/experiments/qwen3/model/state_dict_adapter.py]: Qwen3 don't need to dequantize, because the weights  for Qwen3 are not quantized

File: torchtitan/components/checkpoint.py [hunk 2]
```diff
@@ -709,7 +723,8 @@ def _states_to_load(self, model_only: bool) -> dict[str, Any]:
         """
         # For the first step, we will only load the model.
         if model_only:
-            return self.states[MODEL].state_dict()
+            # Use clean_state_dict for HF loading to avoid unpicklable objects
```
Comment (wwwjn) [inline torchtitan/components/checkpoint.py]: Do we know which fqn key causing the "unpickable error"? We should not need this .

File: torchtitan/models/deepseek_v3/model/quantization.py
Comment (wwwjn) [inline torchtitan/models/deepseek_v3/model/quantization.py:1]: We don't need to touch dequantizd , only deepseek model weights are quantized on huggingface, see : https://huggingface.co/deepseek-ai/DeepSeek-V3-Base/blob/main/README_WEIGHTS.md#dequantization-method. For Qwen3 model, the weights are not quantized (the weight type are BF16 or Float32).

File: torchtitan/components/checkpoint.py (hunk 2 above)
Comment (shuhuayu) [reply torchtitan/components/checkpoint.py]: Thanks for the comment! I double checked the changes on checkpoint.py and they are indeed unnecessary. The previous loading errors were due to other reasons.

File: torchtitan/models/deepseek_v3/model/quantization.py
Comment (shuhuayu) [reply torchtitan/models/deepseek_v3/model/quantization.py:1]: Thanks for pointing this out! I removed the quantization related steps for qwen3.

File: torchtitan/protocols/state_dict_adapter.py [hunk 3]
```diff
@@ -86,3 +93,373 @@ def __init__(
...
+        sorted_experts = [experts[i] for i in sorted_expert_ids]
+        stacked_tensor = torch.stack(sorted_experts, dim=0)
+
+        del expert_weights_by_layer[layer_num][abstract_key]
+        if not expert_weights_by_layer[layer_num]:
+            del expert_weights_by_layer[layer_num]
+
+        return stacked_tensor
+
+    def _dequantize(self, state_dict: dict[str, Any]) -> dict[str, Any]:
```
Comment (wwwjn) [inline torchtitan/protocols/state_dict_adapter.py]: I would suggest _dequantize() and `_add_quantization_scale_inv_tensors()` being put into dsv3 state_dict_adapter, instead of putting it here.

**Comment by wwwjn** [comment]
Thanks for working on this again! Can you attach a screenshot of your local run after loading HF weights?

File: torchtitan/protocols/state_dict_adapter.py (hunk 3 above)
Comment (shuhuayu) [reply torchtitan/protocols/state_dict_adapter.py]: Great suggestion! I have put these quantization related methods back to the deepseek_v3 folder as they are for now specific to deepseek models.

**Comment by shuhuayu** [comment]
> Thanks for working on this again! Can you attach a screenshot of your local run after loading HF weights?

Sure, here is a local run after loading weights from huggingface Qwen/Qwen3-30B-A3B-Instruct-2507 model:

image

File: torchtitan/protocols/state_dict_adapter.py [hunk 4]
```diff
@@ -86,3 +91,323 @@ def __init__(
                     self.fqn_to_index_mapping[hf_key] = int(indx)
             else:
                 self.fqn_to_index_mapping = None
+
+    def _calculate_strided_shard_shard_indices(
```
Comment (tianyu-l) [inline torchtitan/protocols/state_dict_adapter.py]: These util functions shouldn't be put into the `protocol` folder. In particular, they shouldn't be shared by all subclasses / instances.
You could consider one of the following:
1. Still put it under `dsv3/model/state_dict_adapter.py`, and let the qwen3 adapter subclass the dsv3 adapter. (logically not sound but easier to make it work for now)
2. put the code in `models/utils.py`, either by refactoring them into static functions, or create a subclass `MoEStateDictAdapter` (and let DSv3 / Qwen3 sd_adpater subclass it).

**Review by tianyu-l — CHANGES_REQUESTED** [review]

File: torchtitan/protocols/state_dict_adapter.py (hunk 4 above)
Comment (shuhuayu) [reply torchtitan/protocols/state_dict_adapter.py]: Thank you for the suggestion. I have moved the code into `models/utils.py` and created a subclass, `MoEStateDictAdapter`, which can be inherited by both the dsv3 and qwen3 `state_dict_adapter.py` modules.

File: torchtitan/components/checkpoint.py [hunk 5]
```diff
@@ -709,6 +709,7 @@ def _states_to_load(self, model_only: bool) -> dict[str, Any]:
         """
         # For the first step, we will only load the model.
         if model_only:
+
```
Comment (tianyu-l) [inline torchtitan/components/checkpoint.py]: Is this change intentional? If not let's revert.

File: torchtitan/models/utils.py [hunk 6]
```diff
@@ -0,0 +1,349 @@
...
+        )
+        if len(experts) < expected_n_experts:
+            return None
+
+        sorted_expert_ids = sorted(experts.keys())
+        sorted_experts = [experts[i] for i in sorted_expert_ids]
+        local_tensor = torch.stack(sorted_experts, dim=0)._local_tensor
+
+        assert (
+            abstract_key in self.grouped_expert_weight_placements
```
Comment (tianyu-l) [inline torchtitan/models/utils.py:290]: It seems `self.grouped_expert_weight_placements` is not available in this class?

File: torchtitan/models/deepseek_v3/model/state_dict_adapter.py [hunk 7]
```diff
@@ -78,326 +76,6 @@ def __init__(
         self.grouped_expert_weight_shape = {}  # {titan_abstract_key: shape}
         self.local_experts_indices = {}  # {titan_abstract_key: (start_idx, end_idx)}
```
Comment (tianyu-l) [inline torchtitan/models/deepseek_v3/model/state_dict_adapter.py]: This is defined in the parent class. Do we still need to define it here?
The same for `self.model_args`.

File: torchtitan/models/utils.py [hunk 8]
```diff
@@ -0,0 +1,349 @@
...
+from torch.distributed.tensor import DTensor
+from torch.distributed.tensor.placement_types import _StridedShard, Replicate, Shard
+
+from torchtitan.protocols.model import BaseModelArgs
+from torchtitan.protocols.state_dict_adapter import StateDictAdapter
+
+
+class MoEStateDictAdapter(StateDictAdapter):
+    """
+    StateDictAdapter for MoE models.
```
Comment (tianyu-l) [inline torchtitan/models/utils.py:21]: It would be good if we could explain why having such methods.
At a high level, it's because HF models store experts as a module list each with 2D weights. In torchtitan, we store experts as a 3D param with the first dimension being `num_experts`.
The functions in this class helps convert 3D param into list of 2D params so that the checkpoint can be loaded without incurring local memory overhead, and then concatenate the results back to 3D param.
@wwwjn please correct me if I stated it wrong.

**Comment by vwxyzjn** [comment]
@shuhuayu nice PR! The loss and grad norms looks a bit high though -- any idea why? 

Another good way to validate the implementation is to run inference to check if you can get the same output tokens as the HF implementaiton.

File: torchtitan/experiments/qwen3/model/model.py [hunk 9]
```diff
@@ -17,6 +17,7 @@
 
 from .args import Qwen3ModelArgs
 
+
```
Comment (wwwjn) [inline torchtitan/experiments/qwen3/model/model.py]: Revert

File: torchtitan/models/deepseek_v3/model/quantization.py [hunk 10]
```diff
@@ -5,6 +5,7 @@
 # LICENSE file in the root directory of this source tree.
 
 import torch
+
```
Comment (wwwjn) [inline torchtitan/models/deepseek_v3/model/quantization.py]: revert

**Comment by shuhuayu** [comment]
> @shuhuayu nice PR! The loss and grad norms looks a bit high though -- any idea why?
> 
> Another good way to validate the implementation is to run inference to check if you can get the same output tokens as the HF implementaiton.

@vwxyzjn Thanks for the suggestion! I guess the large losses and gradient norms may be due to the suboptimal training configs we set for debug. We have verified the [KL Divergence between a huggingface model and a converted torchtitan model](https://github.com/pytorch/torchtitan/blob/main/scripts/checkpoint_conversion/README.md) on Qwen3 30B-A3B.

test

**Review by tianyu-l — APPROVED** [review]
LGTM
In the future, we should consider adding unit tests for `MoEStateDictAdapter`.

File: torchtitan/experiments/qwen3/model/state_dict_adapter.py [hunk 11]
```diff
@@ -15,72 +15,152 @@
...
+                    )
+                    hf_state_dict.update(local_expert_fqn)
+
+                else:
+                    # keep this path for offline conversion
+                    split_values = self._split_experts_weights(
+                        value, self.model_args.moe_args.num_experts
+                    )
+
+                    for expert_num in range(0, self.model_args.moe_args.num_experts):
```
Comment (fegin) [inline torchtitan/experiments/qwen3/model/state_dict_adapter.py]: nit:

```suggestion
                    for expert_num in range(self.model_args.moe_args.num_experts):
```

File: torchtitan/models/utils.py [hunk 12]
```diff
@@ -4,12 +4,362 @@
...
+            local_expert_tensors[expert_key] = expert_dtensor
+
+        return local_expert_tensors
+
+    def _concatenate_expert_weights_dtensor(
+        self,
+        expert_weights_by_layer: dict[str, dict[str, dict[int, torch.Tensor]]],
+        abstract_key: str,
+        layer_num: str,
+        device_mesh: "DeviceMesh",
```
Comment (fegin) [inline torchtitan/models/utils.py]: Curious why "DeviceMesh" instead of DeviceMesh? DeviceMesh is already imported and defined.

File: torchtitan/models/utils.py [hunk 13]
```diff
@@ -4,12 +4,362 @@
...
+
+        return start_index, end_index
+
+    def _caculate_indices_from_placements(
+        self,
+        dim: int,
+        dim_size: int,
+        dtensor_placements: tuple,
+        device_mesh: "DeviceMesh",
+    ):
```
Comment (fegin) [inline torchtitan/models/utils.py]: Can we have the return typing? `tuple[int, int]`?

Comment (shuhuayu) [reply torchtitan/models/utils.py]: Suggestions incorporated, thank you for the feedback!
